### ISSUE TITLE
Two questions about Functionally Equivalent Extraction

### ISSUE THREAD

**Issue by puro2xz** [IBM/adversarial-robustness-toolbox#420]
When running the function -- _critical_point_search，the program is stuck in an infinite loop. It seems that **the conditions for jumping out of the loop can never be met.** 
The variable delta has been getting smaller and finally became 0.0 but the critical point were not found.  

![TIM图片20200519232358](https://user-images.githubusercontent.com/38450887/82345437-d79cc900-9a27-11ea-9d8d-8284961d242a.png)
  
The first value is delta after running a few seconds. The second and third are m_1, m
_2.  I can't get the result in the end. Can you help me？Thanks a lot!

Another small problem is that if I want to run this code again, I need to delete the original model, otherwise I will get an error.

 - OS: windows 10
 - Python version 3.73
 - ART version or commit number new
 - TensorFlow 1.14.0 / Keras  2.3.1/ PyTorch / MXNet version

**Comment by beat-buesser** [comment]
Hi @puro2xz Thank you very much for using ART and its extraction tools! The behaviour you have observed could be possible, although we definitely want to avoid it. Would you have a script with a seed set that reproduces the case you are describing? I would be very interested to take a look.

**Comment by xiaoxianghuayu** [comment]
I also encountered this problem. I didn't change the `functionally_equivalent_extraction.py` and just run the code. I tried to change the default seed such as 20200520, 20200521, but also never jump out the loop on line 205. 
I printed some variables: 

t_current:  -255.9990966796875
t_current:  -255.99870605468752
t_current:  -255.99831542968752
t_current:  -255.98581542968753
t_current:  -255.98542480468754
t_current:  -255.98542480468755

and after that 

delta:  5.684341886080802e-15
delta:  2.842170943040401e-15
delta:  1.4210854715202005e-15
delta:  7.105427357601002e-16
delta:  3.552713678800501e-16
delta:  1.7763568394002506e-16
...
until zero, and then loop.
between those outputs, there will have such tips
![1](https://user-images.githubusercontent.com/43427419/82474662-f3729e80-9afd-11ea-9087-36659ed0e1f0.png)

I am not sure what is the seed set... but this is my situation.
Python 3.7.0
Keras                                2.3.1
Keras-Applications           1.0.8
Keras-Preprocessing        1.1.0
scikit-image                      0.16.2
scikit-learn                        0.22.1
numpy                              1.16.0
tensorflow                        1.14.0
tensorflow-estimator       1.14.0
tensorflow-gpu                1.14.0

Thanks for all your reading.

**Comment by beat-buesser** [comment]
@xiaoxianghuayu Thank you very much!

@puro2xz I think the error about loading the model can occur for TensorFlow 1.14, it seems to disappear for TensorFlow 1.15.2 and later.

I also have run `functionally_equivalent_extraction.py` and I observe the same behaviour. It looks like something must have changed in one of the dependencies with regards to casting floating-point formats. I think the script enters the infinite loop because of reduced accuracy. If I first delete any old model and train and use a new model in `float64` the infinite loop is avoided as before, e.g. by adding the third line below:
```
import tensorflow as tf

tf.compat.v1.disable_eager_execution()
tf.keras.backend.set_floatx('float64')
```
For me this leads to a case where the script does not find all critical points, therefore I tried a different seed and `np.random.seed(1)` works for me. I will add these updates to ART's next release. Please let me know if this solution works for you.

I think it would be interesting to investigate if it would be possible to make this extraction attack work with single precision (float32).

**Comment by puro2xz** [comment]
@beat-buesser Thank you for helping me solve this problem! The small error disappeared for TensorFlow 1.15.2. Your suggestion is very effective.

After adding the third line you mentioned , I also tried a different seed and `np.random.seed(1)` did not work for me but `np.random.seed(905)` did. It is just a random number.I guess different environments require different random seed.

**Comment by Dainiuss1** [comment]
Hey,

Currently facing same problem as @puro2xz and @xiaoxianghuayu when delta converges to 0 and causes infinite loop. Already tried to use np.random.seed(905), np.random.seed(1), however these changes didn't fix my problem. 
Maybe have any thoughts how can I solve this situation?

Thanks in advance!

Tensorflow==1.15.2
Keras==2.3.1
ART==1.5.1
